resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/usecase/IFileTest.java

```diff
@@ -121,45 +116,37 @@ public void testFile() throws CoreException {
 		file.create(createInputStream("0123456789"), false, monitor);
 
 		// Now tests pass that require that the file exists.
-		assertThat("file does not exist: " + file, file.exists());
-		assertThat("no member with name '" + file.getName() + "' exists in folder: " + folder,
-				folder.findMember(file.getName()).exists());
-		assertThat(workspace.getRoot().findMember(file.getFullPath()), is(file));
-		assertThat("no member at path '" + file.getFullPath() + "' exists in workspace",
-				workspace.getRoot().exists(file.getFullPath()));
-		assertThat(file.getLocation(), is(absolutePath));
+		assertThat(file).matches(IFile::exists, "exists")
+				.matches(it -> workspace.getRoot().exists(it.getFullPath()), "is contained in workspace")
+				.matches(it -> folder.findMember(it.getName()).exists(), "is contained in folder: " + folder)
+				.isEqualTo(workspace.getRoot().findMember(file.getFullPath()));
```

Comment:
I have now also replaced several pure boolean assertions with `matches` assertions that improve the output in case of a failure, in particular because they also output the subject under test. This makes it far easier to provide a helpful error message with minimal custom text.

Reply:
That's an interesting technique that I will remember, thanks for that hint. Regarding earlier discussions: You are completely right about as() and withFailureMessage() sometimes not fitting well, therefore in practice it's sometimes more ugly than the nice theory I painted before. Personally I try to use as() as much as possible, if the assertion itself has helpful text (like isNonEmptyDirectory), but switch to withFailMessage, if the default error message is to generic (like isTrue). Since withFailMessage can take format strings and objects, that's also okay performance wise. However, I'll probably use more of your matches(...) version in the future, as that probably simplifies custom error messages a bit.

Reply:
As a side note: a method reference used for matching can even be negated easily with `Predication.not()`, which, when statically imported, can be used like this: `assertThat(file).matches(not(IFile::exists), "does not exist")`.